aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -360,4 +416,28 @@ protected void finalize() throws Throwable {
       LOG.warn("Unclosed output stream created by:\n\t{}", trace);
     }
   }
+
+  private static class FileAndDigest {
```

Comment:
Could we make this extend File or make it more generic (FileReference) because we use this for both when we have a digest and when we don't.

This could simplify the etags checks on upload to just be:

```
If (file.hasDigest()) {
  requestBuilder.contentMD5(...)
}
```

It just feels like we're using the etag flag to drive a lot of logic in different places.

Reply:
We can do that for multi part request, but for put request doing this is a bit awkward as either we will have to check for all files or just rely on the existing etag flag. I will update it to use hasDigest never the less for places where we can use it.